pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -0,0 +1,789 @@
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+	"k8s.io/kubernetes/pkg/scheduler/framework/plugins/feature"
+	"k8s.io/kubernetes/pkg/scheduler/framework/plugins/names"
+	"k8s.io/kubernetes/pkg/util/dynamic-resource-allocation/resourceclaim"
+)
+
+const (
+	// Name is the name of the plugin used in Registry and configurations.
+	Name = names.DynamicResources
+
+	stateKey framework.StateKey = Name
+)
+
+// The state is initialized in PreFilter phase. Because we save the pointer in
+// framework.CycleState, in the later phases we don't need to call Write method
+// to update the value
+type stateData struct {
+	// A copy of all claims for the Pod (i.e. 1:1 match with
+	// pod.Spec.ResourceClaims), initially with the status from the start
+	// of the scheduling cycle. Each claim instance is read-only because it
+	// might come from the informer cache. The instances get replaced when
+	// the plugin itself successfully does an Update.
+	//
+	// Empty if the Pod has no claims.
+	claims []*resourcev1alpha1.ResourceClaim
+
+	// The AvailableOnNodes node filters of the claims converted from the
+	// v1 API to nodeaffinity.NodeSelector by PreFilter for repeated
+	// evaluation in Filter. Nil for claims which don't have it.
+	availableOnNodes []*nodeaffinity.NodeSelector
+
+	// The indices of all claims that:
+	// - are allocated
+	// - use delayed allocation
+	// - were not available on at least one node
+	//
+	// Set in parallel during Filter, so write access there must be
+	// protected by the mutex. Used by PostFilter.
+	unavailableClaims sets.Int
+
+	// A pointer to the PodScheduling object for the pod, if one exists.
+	// Gets set on demand.
+	//
+	// Conceptually, this object belongs into the scheduler framework
+	// where it might get shared by different plugins. But in practice,
+	// it is currently only used by dynamic provisioning and thus
+	// managed entirely here.
+	podScheduling *resourcev1alpha1.PodScheduling
+
+	// podSchedulingDirty is true if the current copy was locally modified.
+	podSchedulingDirty bool
+
+	mutex sync.Mutex
+}
+
+func (d *stateData) Clone() framework.StateData {
+	return d
+}
+
+func (d *stateData) updateClaimStatus(ctx context.Context, clientset kubernetes.Interface, index int, claim *resourcev1alpha1.ResourceClaim) error {
+	// TODO (?): replace with patch operation. Beware that patching must only succeed if the
+	// object has not been modified in parallel by someone else.
+	claim, err := clientset.ResourceV1alpha1().ResourceClaims(claim.Namespace).UpdateStatus(ctx, claim, metav1.UpdateOptions{})
+	// TODO: metric for update results, with the operation ("set selected
+	// node", "set PotentialNodes", etc.) as one dimension.
+	if err != nil {
+		return fmt.Errorf("update resource claim: %w", err)
+	}
+
+	// Remember the new instance. This is relevant when the plugin must
+	// update the same claim multiple times (for example, first set
+	// PotentialNodes, then SelectedNode), because otherwise the second
```

Comment:
Can we do only one API call per scheduling-cycle?

Reply:
The example in the comment is out-dated: in the normal case, setting PotentalNodes and SelectedNode are already combined and there is only a single call per cycle. It wasn't like that initially.

Where it can still occur is when a claim gets reserved and then scheduling cannot complete. I've updated the comment to use that as example.